File: android/SwiftKotlination/modules/test/src/main/kotlin/fr/jhandguy/test/network/mocks/NetworkManagerMock.kt
package fr.jhandguy.test.network.mocks

import fr.jhandguy.network.model.network.NetworkManagerInterface
import fr.jhandguy.network.model.network.Request
import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer
import fr.jhandguy.network.model.observer.Result
import kotlin.collections.ArrayList

class NetworkManagerMock(
    val result: Result<ByteArray>,
    var observers: MutableList<Observer<ByteArray>> = ArrayList()
) : NetworkManagerInterface {

    override suspend fun observe(request: Request, observer: Observer<ByteArray>): Disposable {
        observers.add(observer)
        execute(request)

        return Disposable {}
    }

    override suspend fun execute(request: Request) = observers.forEach { it(result) }
}

File: ios/Modules/TestKit/TestKit/Sources/Mocks/Network/Model/NetworkManagerMock.swift
import Foundation
import NetworkKit

public final class NetworkManagerMock: NetworkManagerProtocol {
    public var result: Result<Data, Error>
    var observers: [Observer<Data>] = []

    public init(result: Result<Data, Error>) {
        self.result = result
    }

    @discardableResult
    public func observe(_ request: Request, _ observer: @escaping Observer<Data>) -> Disposable {
        observers.append(observer)
        execute(request)

        return Disposable {}
    }

    public func execute(_: Request) {
        observers.forEach { $0(result) }
    }
}
